libs/application-generic/src/usecases/merge-preferences/merge-preferences.usecase.ts

```diff
@@ -0,0 +1,189 @@
+ * 2. Workflow user preferences
+ * 3. Subscriber global preferences
+ * 4. Subscriber workflow preferences
+ *
+ * If a workflow has the readOnly flag set to true, the subscriber preferences are ignored.
+ *
+ * If the workflow does not have the readOnly flag set to true, the subscriber preferences are merged with the workflow preferences.
+ *
+ * If the subscriber has no preferences, the workflow preferences are returned.
+ */
+export class MergePreferences {
+  public static merge(
+    command: MergePreferencesCommand,
+  ): GetPreferencesResponseDto {
+    const workflowResourcePreferences = this.getWorkflowResourcePreferences(
+      command.preferences,
+    );
+    const workflowUserPreferences = this.getWorkflowUserPreferences(
+      command.preferences,
+    );
+
+    const workflowPreferences = deepMerge(
+      [workflowResourcePreferences, workflowUserPreferences]
+        .filter((preference) => preference !== undefined)
+        .map((item) => item.preferences),
+    ) as WorkflowPreferences;
+
+    const subscriberGlobalPreferences = this.getSubscriberGlobalPreferences(
+      command.preferences,
+    );
+    const subscriberWorkflowPreferences = this.getSubscriberWorkflowPreferences(
+      command.preferences,
+    );
+
+    const subscriberPreferences = deepMerge(
+      [subscriberGlobalPreferences, subscriberWorkflowPreferences]
+        .filter((preference) => preference !== undefined)
+        .map((item) => item.preferences),
+    );
+
+    /**
+     * Order is important here because we like the workflowPreferences (that comes from the bridge)
+     * to be overridden by any other preferences and then we have preferences defined in dashboard and
+     * then subscribers global preferences and the once that should be used if it says other then anything before it
+     * we use subscribers workflow preferences
+     */
+    const preferencesEntities = [
+      workflowResourcePreferences,
+      workflowUserPreferences,
+      subscriberGlobalPreferences,
+      subscriberWorkflowPreferences,
+    ];
+    const source = Object.values(PreferencesTypeEnum).reduce(
+      (acc, type) => {
+        const preference = command.preferences.find(
+          (item) => item.type === type,
+        );
+        if (preference) {
+          acc[type] = preference.preferences as WorkflowPreferences;
+        } else {
+          acc[type] = null;
+        }
+
+        return acc;
+      },
+      {} as GetPreferencesResponseDto['source'],
+    );
+    const preferences = preferencesEntities
+      .filter((preference) => preference !== undefined)
+      .map((item) => item.preferences);
+
+    // ensure we don't merge on an empty list
+    if (preferences.length === 0) {
+      return { preferences: undefined, type: undefined, source };
+    }
+
+    const readOnlyFlag = workflowPreferences?.all?.readOnly;
+
+    // Determine the most specific preference applied
+    let mostSpecificPreference: PreferencesTypeEnum | undefined;
```

Comment:
We can get this automatically from `deepMerge` if we provide it during function invocation. The last type will win.

Reply:
Yes! Done